### DISCUSSION TITLE
Custom code doesn't compile when using ParallelRunner functions

### DISCUSSION THREAD

**Discussion by KMadura-Solaris** [libjxl/libjxl#3749] [Q&A] [answered]
Hello

I am learning how to use this library in C++ and it all went well until now. I have made a simple program which takes a PNG file name as input, reads this file and successfully creates a new JXL file with lossless compression. Compression however takes some time and my next step was to use multiple threads.

This is a piece of code added right at the beginning:
```c++
        auto runner = JxlThreadParallelRunnerCreate(
            NULL,
            JxlThreadParallelRunnerDefaultNumWorkerThreads()
        );
        if (JxlEncoderSetParallelRunner(jxl_enc, JxlThreadParallelRunner, runner) == JxlEncoderStatus::JXL_ENC_ERROR) {
            error_text = "Could not set parallel runners";
            return false;
        }
```
in between `JxlEncoder* jxl_enc = JxlEncoderCreate(NULL);` and `JxlBasicInfo jxl_basic_info;`

When trying to compile my code I get:
```text
/usr/bin/ld: CMakeFiles/jxlexp.dir/main.cpp.o: in function `KM_jxl::compress_an_image(KM_png&)':
main.cpp:(.text._ZN6KM_jxl17compress_an_imageER6KM_png[_ZN6KM_jxl17compress_an_imageER6KM_png]+0x49): undefined reference to `JxlThreadParallelRunnerDefaultNumWorkerThreads'
/usr/bin/ld: main.cpp:(.text._ZN6KM_jxl17compress_an_imageER6KM_png[_ZN6KM_jxl17compress_an_imageER6KM_png]+0x56): undefined reference to `JxlThreadParallelRunnerCreate'
/usr/bin/ld: main.cpp:(.text._ZN6KM_jxl17compress_an_imageER6KM_png[_ZN6KM_jxl17compress_an_imageER6KM_png]+0x72): undefined reference to `JxlThreadParallelRunner'
```

which means those functions are probably missing in final libjxl.

I am not using default `libjxl-dev` and `libjxl0.7` supplied by my OS repository (Ubuntu 23.10 on WSL and those old packages are not installed at the moment). Instead I built libjxl from v0.10.x repo according to `BUILDING.md` instruction.

Everything else works absolutely fine except creating and managing parallel runners. What am I doing wrong? Is this functionality removed/deprecated?

**Comment by veluca93** [comment, marked as answer]
The parallel runners are included in libjxl_threads, not in the main jxl library - are you linking that in?

**Comment by KMadura-Solaris** [comment]
Thank you so much! jxl_threads was missing in CMakeLists.txt and I was running in circles trying to find an issue in wrong place.
